sdk/core/azure-core/azure/core/messaging.py

```diff
@@ -0,0 +1,145 @@
+     and must not exceed the length of 20 characters.
+    :type extensions: Optional[dict]
+    :ivar source: Identifies the context in which an event happened. The combination of id and source must
+     be unique for each distinct event. If publishing to a domain topic, source must be the domain name.
+    :vartype source: str
+    :ivar data: Event data specific to the event type.
+    :vartype data: object
+    :ivar type: Type of event related to the originating occurrence.
+    :vartype type: str
+    :ivar time: The time (in UTC) the event was generated.
+    :vartype time: ~datetime.datetime
+    :ivar dataschema: Identifies the schema that data adheres to.
+    :vartype dataschema: str
+    :ivar datacontenttype: Content type of data value.
+    :vartype datacontenttype: str
+    :ivar subject: This describes the subject of the event in the context of the event producer
+     (identified by source).
+    :vartype subject: str
+    :ivar specversion: Optional. The version of the CloudEvent spec. Defaults to "1.0"
+    :vartype specversion: str
+    :ivar id: An identifier for the event. The combination of id and source must be
+     unique for each distinct event. If not provided, a random UUID will be generated and used.
+    :vartype id: str
+    :ivar extensions: A CloudEvent MAY include any number of additional context attributes
+     with distinct names represented as key - value pairs. Each extension must be alphanumeric, lower cased
+     and must not exceed the length of 20 characters.
+    :vartype extensions: dict
+    """
+    def __init__(self, source, type, **kwargs): # pylint: disable=redefined-builtin
+        # type: (str, str, Any) -> None
+        self.source = source
+        self.type = type
+        self.specversion = kwargs.pop("specversion", "1.0")
+        self.id = kwargs.pop("id", str(uuid.uuid4()))
+        self.time = kwargs.pop("time", datetime.now(TZ_UTC))
+        self.datacontenttype = kwargs.pop("datacontenttype", None)
+        self.dataschema = kwargs.pop("dataschema", None)
+        self.subject = kwargs.pop("subject", None)
+        self.extensions = {}
+        _extensions = dict(kwargs.pop('extensions', {}))
+        for key in _extensions.keys():
+            if not key.islower() or not key.isalnum():
+                raise ValueError("Extension attributes should be lower cased and alphanumeric.")
+        self.extensions.update(_extensions)
+        self.data = kwargs.pop("data", None)
+
+    def __repr__(self):
+        return (
+            "CloudEvent(source={}, type={}, specversion={}, id={}, time={})".format(
+                self.source,
+                self.type,
+                self.specversion,
+                self.id,
+                self.time
+            )[:1024]
+        )
+
+    @classmethod
+    def from_dict(cls, event, **kwargs):
+        # type: (Dict, Any) -> CloudEvent
```

Comment:
Type annotation is incorrect - you will return an instance of cls. Off the top of my head, this would be:

`# type: (Type[T], Dict, **Any) -> T`

where T is a typing.TypeVar (ideally bound to `CloudEvent`)